File: src/inoio_sandbox/worktree.py
import hashlib
import shutil
import subprocess
from pathlib import Path

from inoio_sandbox import log


def project_slug() -> str:
    """Return a stable short slug for the current repo."""
    cwd = Path.cwd()
    git_common_dir = _git_common_dir(cwd)
    if git_common_dir:
        key = str(git_common_dir.resolve())
    else:
        log.warn("not inside a git repo; using CWD hash as project slug.")
        key = str(cwd.resolve())
    h = hashlib.sha256(key.encode()).hexdigest()[:8]
    return f"p-{h}"


def branch_slug(branch: str) -> str:
    """Sanitize a branch name for use in VM names (slashes are invalid)."""
    return branch.replace("/", "-")


def _git_common_dir(cwd: Path) -> Path | None:
    try:
        output = subprocess.check_output(
            ["git", "rev-parse", "--git-common-dir"],
            cwd=cwd,
            stderr=subprocess.DEVNULL,
        )
        path = Path(output.decode().strip())
        if not path.is_absolute():
            path = cwd / path
        return path.resolve()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None


def branch_name(cwd: Path | None = None) -> str:
    cwd = cwd or Path.cwd()
    try:
        output = subprocess.check_output(
            ["git", "rev-parse", "--abbrev-ref", "HEAD"],
            cwd=cwd,
            stderr=subprocess.DEVNULL,
        )
    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
        raise RuntimeError(f"Unable to determine current git branch from {cwd}") from exc
    return output.decode().strip()


def current_worktree_path(cwd: Path | None = None) -> Path | None:
    cwd = cwd or Path.cwd()
    try:
        output = subprocess.check_output(
            ["git", "rev-parse", "--show-toplevel"],
            cwd=cwd,
            stderr=subprocess.DEVNULL,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return None
    return Path(output.decode().strip()).resolve()


def worktree_path(state_dir: Path, project_slug: str, branch: str) -> Path:
    return state_dir / "worktrees" / project_slug / branch


def _is_git_worktree(path: Path) -> bool:
    try:
        subprocess.check_output(
            ["git", "rev-parse", "--is-inside-work-tree"],
            cwd=path,
            stderr=subprocess.DEVNULL,
        )
    except (subprocess.CalledProcessError, FileNotFoundError):
        return False
    return True


def ensure_worktree(repo_root: Path, state_dir: Path, project_slug: str, branch: str) -> Path:
    target = worktree_path(state_dir, project_slug, branch)
    if target.exists():
        if _is_git_worktree(target):
            return target
        shutil.rmtree(target)
    target.parent.mkdir(parents=True, exist_ok=True)
    subprocess.run(
        ["git", "worktree", "add", str(target), branch],
        cwd=repo_root,
        check=True,
    )
    return target

File: internal/opencodemsb/worktree.go
package opencodemsb

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

func ProjectSlug() string {
	commonDir, err := gitCommonDir(".")
	if err != nil || commonDir == "" {
		cwd, _ := filepath.Abs(".")
		warn("not inside a git repo; using CWD hash as project slug.")
		h := sha256.Sum256([]byte(cwd))
		return "p-" + hex.EncodeToString(h[:])[:8]
	}
	abs, _ := filepath.Abs(commonDir)
	h := sha256.Sum256([]byte(abs))
	return "p-" + hex.EncodeToString(h[:])[:8]
}

func BranchSlug(branch string) string {
	return strings.ReplaceAll(branch, "/", "-")
}

func gitCommonDir(cwd string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--git-common-dir")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	p := strings.TrimSpace(string(out))
	if !filepath.IsAbs(p) {
		p = filepath.Join(cwd, p)
	}
	return filepath.Abs(p)
}

func BranchName(cwd string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--abbrev-ref", "HEAD")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("unable to determine current git branch from %s: %w", cwd, err)
	}
	return strings.TrimSpace(string(out)), nil
}

func CurrentWorktreePath(cwd string) (string, error) {
	cmd := exec.Command("git", "rev-parse", "--show-toplevel")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return filepath.Abs(strings.TrimSpace(string(out)))
}

func WorktreePath(stateDir, projectSlug, branch string) string {
	return filepath.Join(stateDir, "worktrees", projectSlug, branch)
}

func isGitWorktree(path string) bool {
	cmd := exec.Command("git", "rev-parse", "--is-inside-work-tree")
	cmd.Dir = path
	return cmd.Run() == nil
}

func EnsureWorktree(repoRoot, stateDir, projectSlug, branch string) (string, error) {
	target := WorktreePath(stateDir, projectSlug, BranchSlug(branch))
	if info, err := os.Stat(target); err == nil && info.IsDir() {
		if isGitWorktree(target) {
			return target, nil
		}
		os.RemoveAll(target)
	}
	if err := os.MkdirAll(filepath.Dir(target), 0o755); err != nil {
		return "", fmt.Errorf("create worktree parent dir: %w", err)
	}
	cmd := exec.Command("git", "worktree", "add", target, branch)
	cmd.Dir = repoRoot
	if out, err := cmd.CombinedOutput(); err != nil {
		return "", fmt.Errorf("git worktree add failed: %w: %s", err, string(out))
	}
	return target, nil
}

func RemoveWorktree(path string) error {
	cmd := exec.Command("git", "worktree", "remove", "--force", path)
	return cmd.Run()
}
